crates/red_knot/src/module.rs

```diff
@@ -0,0 +1,708 @@
+#![allow(unreachable_pub)]
+
+use std::fmt::Formatter;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::AtomicU32;
+use std::sync::Arc;
+
+use crate::files::{FileId, Files};
+use crate::FxDashMap;
+use dashmap::mapref::entry::Entry;
+
+/// ID uniquely identifying a module.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct ModuleId(u32);
+
+/// A module name, e.g. `foo.bar`.
+///
+/// Always normalized to the absolute form (never a relative module name).
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct ModuleName(smol_str::SmolStr);
```

Comment:
I thought about this as well and don't have a final conclusion. The good thing, it's nicely abstracted away, so we can play with it when we have real world usage. 

My thinking why I left it as is:

* We could use a `SmallVec` with a size of 2-3 in combination with a `SmolStr`. That means we should avoid allocating in many cases. However, it does have the downside that `ModuleName` now becomes a 3 * 24 bytes struct. That's probably not too bad because it is not used in many places but something to consider. 
* The worst case is that we need 1 + n where `n` is the number of segments allocations
* I think splitting by `.` is cheap. `PathBuf` uses a single vector internally and `.components` splits on demand. 

I went with the string because it felt simpler for now and I don't think splitting by components is a very common (hot) operation.